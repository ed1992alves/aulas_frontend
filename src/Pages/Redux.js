import React, { useRef, useEffect, useState } from "react";
import {
  Grey,
  Orange,
  Green,
  Purple,
  Blue,
  Yellow
} from "../Components/Text/Text";

export const Redux = () => {
  return (
    <React.Fragment>
      <h3>Redux</h3>
      <h2>Actions</h2>
      <p>
        Actions are payloads of information that send data from your application
        to your store. They are the only source of information for the store.
        You send them to the store using store.dispatch().
      </p>
      <div className="examples scopes">
        <div>{"{"}</div>
        <div className="nr_idents1">
          type: <Purple>ADD_TODO</Purple>,
        </div>
        <div className="nr_idents1">
          {" "}
          text: <Green>'Build my first Redux app'</Green>
        </div>
        <div>{"}"}</div>
      </div>
      <p>
        Types should typically be defined as string constants. Once your app is
        large enough, you may want to move them into a separate module.
      </p>
      <h3>Action Creators</h3>
      <p>Action creators are exactly that—functions that create actions. </p>
      <div className="examples scopes">
        <div>
          <Purple>function</Purple> <Orange>addTodo</Orange> (text) {"{"}
        </div>
        <div className="nr_idents1"> return {"{"}</div>
        <div className="nr_idents2">
          type: <Purple>ADD_TODO</Purple>,
        </div>
        <div className="nr_idents2">text</div>
        <div className="nr_idents1">{"}"}</div>
        <div>{"}"}</div>
      </div>
      <p>This makes them portable and easy to test.</p>

      <h2>Dispatch</h2>
      <p>
        Dispatches an action. This is the only way to trigger a state change.
      </p>
      <div className="examples scopes">
        <div>
          <Yellow>dispatch</Yellow>(<Yellow>addTodo</Yellow>(text))
        </div>
      </div>
      <h2>Reducer</h2>
      <p>
        Reducers specify how the application's state changes in response to
        actions sent to the store. Remember that actions only describe what
        happened, but don't describe how the application's state changes.
      </p>
      <p>
        The reducer is a pure function that takes the previous state and an
        action, and returns the next state. Things you should never do inside a
        reducer:
        <ul>
          <li>Mutate its arguments;</li>
          <li>Perform side effects like API calls and routing transitions;</li>
          <li>Call non-pure functions, e.g. Date.now() or Math.random().</li>
        </ul>
      </p>
      <div className="examples scopes">
        <div>
          <Purple>function</Purple> <Yellow>todoApp</Yellow>(state =
          initialState, action) {"{"}
        </div>
        <div className="nr_idents1">
          <Purple>switch</Purple> (action.type) {"{"}
        </div>
        <div className="nr_idents2">
          <Purple> case </Purple> <Yellow>SET_VISIBILITY_FILTER</Yellow>:
        </div>
        <div className="nr_idents2">
          <Purple>return</Purple> {"{"}{" "}
        </div>
        <div className="nr_idents3"> ...state, </div>
        <div className="nr_idents3"> visibilityFilter: action.filter </div>
        <div className="nr_idents2">{"}"})</div>
        <div className="nr_idents2">
          <Purple>default:</Purple>
        </div>
        <div className="nr_idents3">
          {" "}
          <Purple>return</Purple> state
        </div>
        <div className="nr_idents2">{"}"}</div>
        <div className="nr_idents1">{"}"}</div>
      </div>
      <h2>Store</h2>
      <p>
        The Store is the object that brings them together. The store has the
        following responsibilities:
      </p>
      <ul>
        <li>Holds application state;</li>
        <li>Allows access to state via getState();</li>
        <li>Allows state to be updated via dispatch(action);</li>
        <li>Registers listeners via subscribe(listener);</li>
      </ul>

      <div className="examples scopes">
        <div>
          <Purple>const</Purple> store = <Yellow>createStore</Yellow>(todoApp,
          window.<Purple>STATE_FROM_SERVER</Purple>)
        </div>
      </div>
      <p>
        {" "}
        You may optionally specify the initial state as the second argument to
        createStore()
      </p>
      <h2>Usage with React</h2>
      <h3>Provider</h3>
      <p>
        The {"<Provider /> "}makes the Redux store available to any nested
        components that have been wrapped in the connect() function. Since any
        React component in a React Redux app can be connected, most applications
        will render a {"<Provider>"} at the top level, with the entire app’s
        component tree inside of it.
      </p>
      <div className="examples scopes">
        <div>
          <Purple>const</Purple> store = createStore()
        </div>
        <p></p>
        <div>
          ReactDOM.<Yellow>render</Yellow>(
        </div>
        <div className="nr_idents1">
          {"<Provider store={"}
          <Green>store</Green>
          {"}>"}
        </div>
        <div className="nr_idents1">{"<App />"}</div>
        <div className="nr_idents1">{"</Provider>,"}</div>
        <div className="nr_idents1">
          <Green>document</Green>.getElementById(<Green>'root'</Green>)
        </div>
        )
      </div>
      <h3>Connect</h3>
      <p>The connect() function connects a React component to a Redux store.</p>
      <p>
        It provides its connected component with the pieces of the data it needs
        from the store, and the functions it can use to dispatch actions to the
        store.
      </p>
      <div className="examples scopes">
        {" "}
        <div>
          <Purple>function</Purple> <Orange>connect</Orange>(mapStateToProps?,
          mapDispatchToProps?, mergeProps?, options?)
        </div>{" "}
      </div>
      <p>
        The mapStateToProps and mapDispatchToProps deals with your Redux store’s
        state and dispatch, respectively.
      </p>
      <h4>mapStateToProps</h4>
      <p>
        If a mapStateToProps function is specified, the new wrapper component
        will subscribe to Redux store updates. This means that any time the
        store is updated, mapStateToProps will be called.
      </p>
      <p>Parameters</p>
      <ul>
        {" "}
        <li>state</li>
        <li>ownProps</li>
      </ul>
      <div class="examples scopes">
        <div>
          <Purple>const</Purple> mapStateToProps = state => ({"{"} todos:
          state.todos {"}"})
        </div>
      </div>
      <h4>mapDispatchToProps</h4>
      <p>Parameters</p>
      <ul>
        {" "}
        <li>dispatch</li>
        <li>ownProps</li>
      </ul>
      <div className="examples scopes">
        <div>
          <Purple>const</Purple> mapDispatchToProps = dispatch => {"{"}
        </div>
        <div className="nr_idents1">
          <Purple>return</Purple> {"{"}
        </div>
        <div className="nr_idents2">
          increment: () => dispatch({"{"} type: <Green>'INCREMENT'</Green> {"}"}
          )
        </div>
        <div className="nr_idents1">{"}"}</div>
        <div>{"}"}</div>
      </div>
      <h2>Selectors</h2>
      <p>
        A “selector” is simply a function that accepts Redux state as an
        argument and returns data that is derived from that state.{" "}
      </p>
      <div className="scopes examples">
        <div>selectUserIds = state => state.users.map(user => user.id);</div>
      </div>
      <h2>React-Redux Hooks</h2>
      <p>
        React Redux now offers a set of hook APIs as an alternative to the
        existing connect() Higher Order Component. These APIs allow you to
        subscribe to the Redux store and dispatch actions, without having to
        wrap your components in connect()
      </p>
      <h3>useSelector()</h3>
      <p>
        Allows you to extract data from the Redux store state, using a selector
        function.
      </p>
      <div class="examples scopes">
        <div>
          <Purple>const</Purple> result: any = useSelector(selector:{" "}
          <Green>Function</Green>)
        </div>
      </div>
      <p>
        The selector is approximately equivalent to the mapStateToProps. The
        selector will be called with the entire Redux store state as its only
        argument. The selector will be run whenever the function component
        renders. useSelector() will also subscribe to the Redux store, and run
        your selector whenever an action is dispatched.
      </p>
      <div class="examples scopes">
        <div>
          <Purple>import</Purple> React from <Green>'react'</Green>
        </div>
        <div>
          <Purple>import</Purple> {"{ useSelector }"} from{" "}
          <Green>'react-redux'</Green>
        </div>
        <div>
          <Purple> export const</Purple> CounterComponent = () => {"{"}{" "}
        </div>
        <div className="nr_idents1">
          {" "}
          const counter = <Orange>useSelector</Orange>(state => state.counter){" "}
        </div>
        <div className="nr_idents1">return {"<div>{counter}</div>"}</div>
        <div>{"}"}</div>
      </div>
      <h3>useDispatch()</h3>
      <div class="examples scopes">
        <div>
          <Purple>const</Purple> dispatch = useDispatch()
        </div>
      </div>
      <p>
        This hook returns a reference to the dispatch function from the Redux
        store. You may use it to dispatch actions as needed.
      </p>
      <div class="examples scopes">
        <div>
          <Purple>import</Purple> React from <Green>'react'</Green>
        </div>
        <div>
          <Purple>import</Purple> {"{ useDispatch }"} from{" "}
          <Green>'react-redux'</Green>
        </div>
        <div>
          {" "}
          <Purple>export const</Purple> CounterComponent = {"({ value }) => {"}{" "}
        </div>
        <div className="nr_idents1">
          {" "}
          <Purple>const</Purple> dispatch = <Orange>useDispatch</Orange>()
        </div>
        <div className="nr_idents1">
          {"return <"}
          <Purple>button</Purple>
          {" onClick={() => "}
          <Orange>dispatch</Orange>
          {"({ type: 'increment-counter' })}>Increment<"}
          <Purple>/button></Purple>
        </div>
        {"}"}
      </div>
    </React.Fragment>
  );
};
