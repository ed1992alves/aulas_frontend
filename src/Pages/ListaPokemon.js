import React, { useState, useRef, useEffect, useReducer } from "react";
import { Dropdown, Card } from "../Components/index";
import { addPokemonList, navigation, setType } from "../Redux/actions";
import { connect } from "react-redux";
import { getPokemonsOnList, getCurrentPage } from "../Redux/selectors";
import "../Style/card.less";
import pikachu from "../Style/images/pokemon/pikachu.png";
import blastoise from "../Style/images/pokemon/blastoise.png";
import charizard from "../Style/images/pokemon/charizard.png";
import flareon from "../Style/images/pokemon/flareon.png";
import vileplume from "../Style/images/pokemon/vileplume.png";
import bulbasaur from "../Style/images/pokemon/bulbasaur.png";
import gengar from "../Style/images/pokemon/gengar.png";
import alakazam from "../Style/images/pokemon/alakazam.png";
import pokeball from "../Style/images/pokemon/pokeball.png";
import { useSelector, useDispatch } from "react-redux";

export const ListaPokemon = ({ startId, endId }) => {
  const dispatch = useDispatch();
  const pokemonsList = useSelector(getPokemonsOnList);

  const typeList = [
    "All",
    "Normal",
    "Poison",
    "Electric",
    "Fire",
    "Water",
    "Rock",
    "Grass",
    "Fighting",
    "Psychic",
    "Ground",
    "Fairy"
  ];

  function getType(type, pokemonName) {
    switch (type) {
      case "speed":
        return pokemonsList[pokemonName]?.stats[0]?.base_stat;
      case "specialDefense":
        return pokemonsList[pokemonName]?.stats[1]?.base_stat;
      case "specialAttack":
        return pokemonsList[pokemonName]?.stats[2]?.base_stat;
      case "defense":
        return pokemonsList[pokemonName]?.stats[3]?.base_stat;
      case "attack":
        return pokemonsList[pokemonName]?.stats[4]?.base_stat;
      case "hp":
        return pokemonsList[pokemonName]?.stats[5]?.base_stat;
      case "weight":
        return pokemonsList[pokemonName]?.weight;
      case "height":
        return pokemonsList[pokemonName]?.height;
      case "type":
        return pokemonsList[pokemonName]?.types[
          pokemonsList[pokemonName]?.types.length - 1
        ]?.type?.name;
      default:
        return;
    }
  }

  async function fetchPokemon(startingId, finalId) {
    let arrayPokemon = [];
    console.log("Start", startingId);
    for (let aux = startingId; aux <= finalId; aux++) {
      await fetch(`https://pokeapi.co/api/v2/pokemon/${aux}`)
        .then(async res => res.json())
        .then(async res => {
          arrayPokemon = { ...arrayPokemon, [res.name]: res };
        })
        .catch(error => console.log(error));
    }

    dispatch(addPokemonList(arrayPokemon, startingId, finalId));
    console.log("End");
  }

  function getAvailabity(type, pokemonName) {
    const ability = pokemonsList[pokemonName]?.abilities?.find(
      ability =>
        (type == "Normal" && !ability.is_hidden) ||
        (type == "Hidden" && ability.is_hidden)
    );

    return ability?.ability?.name;
  }

  const imageSrcSelector = pokemonName => {
    switch (pokemonsList[pokemonName]?.name) {
      case "pikachu":
        return pikachu;
      case "blastoise":
        return blastoise;
      case "charizard":
        return charizard;
      case "flareon":
        return flareon;
      case "bulbasaur":
        return bulbasaur;
      case "gengar":
        return gengar;
      case "alakazam":
        return alakazam;
      default:
        return pokemonsList[pokemonName]?.sprites?.front_default || pokeball;
    }
  };

  useEffect(() => {
    if (pokemonsList) return;
    fetchPokemon(startId, endId);
  });

  if (pokemonsList)
    return (
      <div class="listContainer pokemonPageWrapper">
        <Dropdown
          title="pokemon list"
          name="pokemon"
          items={typeList}
          selected={typeList[0]}
          callback={item => {
            dispatch(setType(item));
          }}
        ></Dropdown>
        <div class="flexContainer">
          {Object.keys(pokemonsList).map((pokemonName, key) => (
            <Card
              key={key}
              name={pokemonName}
              imageSrc={imageSrcSelector(pokemonName)}
              info={{
                speed: getType("speed", pokemonName),
                specialDefense: getType("specialDefense", pokemonName),
                specialAttack: getType("specialAttack", pokemonName),
                defense: getType("defense", pokemonName),
                weight: getType("weight", pokemonName),
                height: getType("height", pokemonName),
                type: getType("type", pokemonName),
                attack: getType("attack", pokemonName),
                hp: getType("hp", pokemonName)
              }}
              ability={{
                normal: getAvailabity("Normal", pokemonName),
                hidden: getAvailabity("Hidden", pokemonName)
              }}
            ></Card>
          ))}
        </div>
      </div>
    );
  else return null;
};
