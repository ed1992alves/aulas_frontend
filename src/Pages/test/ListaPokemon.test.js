import { ListaPokemon } from "../ListaPokemon";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStoreWithInitialState } from "../../Redux/store";
import React from "react";
import * as actions from "../../Redux/actions";

jest.spyOn(actions, "setType");

const state = {
  pokemonList: {
    pikachu: {
      id: 15,
      types: [{ type: { name: "electric" } }],
      stats: [
        {
          base_stat: 90,
          effort: 2,
          stat: {
            name: "speed",
            url: "https://pokeapi.co/api/v2/stat/6/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-defense",
            url: "https://pokeapi.co/api/v2/stat/5/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-attack",
            url: "https://pokeapi.co/api/v2/stat/4/"
          }
        },
        {
          base_stat: 40,
          effort: 0,
          stat: {
            name: "defense",
            url: "https://pokeapi.co/api/v2/stat/3/"
          }
        },
        {
          base_stat: 55,
          effort: 0,
          stat: {
            name: "attack",
            url: "https://pokeapi.co/api/v2/stat/2/"
          }
        },
        {
          base_stat: 35,
          effort: 0,
          stat: {
            name: "hp",
            url: "https://pokeapi.co/api/v2/stat/1/"
          }
        }
      ]
    },
    charizard: {
      id: 60,
      types: [{ type: { name: "fire" } }],
      stats: [
        {
          base_stat: 90,
          effort: 2,
          stat: {
            name: "speed",
            url: "https://pokeapi.co/api/v2/stat/6/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-defense",
            url: "https://pokeapi.co/api/v2/stat/5/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-attack",
            url: "https://pokeapi.co/api/v2/stat/4/"
          }
        },
        {
          base_stat: 40,
          effort: 0,
          stat: {
            name: "defense",
            url: "https://pokeapi.co/api/v2/stat/3/"
          }
        },
        {
          base_stat: 55,
          effort: 0,
          stat: {
            name: "attack",
            url: "https://pokeapi.co/api/v2/stat/2/"
          }
        },
        {
          base_stat: 35,
          effort: 0,
          stat: {
            name: "hp",
            url: "https://pokeapi.co/api/v2/stat/1/"
          }
        }
      ]
    }
  },
  list: {
    start: 10,
    end: 80,
    type: "All"
  }
};

const store = createStoreWithInitialState(state);

describe("ListaPokemon", () => {
  it("snapshot", () => {
    const { container } = render(
      <Provider store={store}>
        <ListaPokemon startId={10} endId={20}></ListaPokemon>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("logic", () => {
    const { getByTestId, getAllByTestId, rerender } = render(
      <Provider store={store}>
        <ListaPokemon startId={10} endId={20}></ListaPokemon>
      </Provider>
    );
    expect(getAllByTestId("card").length).toBe(2);
    getByTestId("dropdownButton").click();
    fireEvent.click(getAllByTestId("dropdownButtonOption")[4]);

    expect(actions.setType).toHaveBeenNthCalledWith(1, "Fire");

    rerender(
      <Provider store={store}>
        <ListaPokemon startId={10} endId={20}></ListaPokemon>
      </Provider>
    );
    expect(getAllByTestId("card").length).toBe(1);
  });
});
