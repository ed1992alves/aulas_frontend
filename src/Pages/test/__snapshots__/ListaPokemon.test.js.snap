// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListaPokemon snapshot 1`] = `
<div>
  <div
    class="listContainer pokemonPageWrapper"
  >
    <div
      class="wrapper pokemon"
      data-testid="dropdown"
      id="dropdown"
    >
      <div
        class="dropdown_title"
      >
        pokemon list
      </div>
      <button
        class="dropdown"
        data-testid="dropdownButton"
      >
        <div
          class="selected"
        >
          All
        </div>
        <div
          class="chevron"
        />
      </button>
      <div
        class="dropdown-list"
      >
        <ul>
          <li
            class="selected"
            data-testid="dropdownButtonOption"
          >
            All
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Normal
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Poison
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Electric
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Fire
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Water
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Rock
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Grass
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Fighting
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Psychic
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Ground
          </li>
          <li
            data-testid="dropdownButtonOption"
          >
            Fairy
          </li>
        </ul>
      </div>
    </div>
    <div
      class="flexContainer"
    >
      <div
        class="card u-full-width electric"
        data-testid="card"
      >
        <div
          class="imageWrapper"
        >
          <img
            class="pokemon-image"
            src=""
          />
        </div>
        <div
          class="textWrapper"
        >
          <div
            class="pokemon-type"
          >
            electric
          </div>
          <div
            class="pokemon-name"
          >
            pikachu
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              HP:
            </div>
             
            <div
              class="pokemon-value"
            >
              35
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Attack:
            </div>
             
            <div
              class="pokemon-value"
            >
              55
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Defense:
            </div>
             
            <div
              class="pokemon-value"
            >
              40
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Special Attack:
            </div>
             
            <div
              class="pokemon-value"
            >
              50
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Special Defense:
            </div>
             
            <div
              class="pokemon-value"
            >
              50
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Speed:
            </div>
             
            <div
              class="pokemon-value"
            >
              90 m/s
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Height:
            </div>
             
            <div
              class="pokemon-value"
            >
              undefined feets
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Weight:
            </div>
             
            <div
              class="pokemon-value"
            >
              undefined pounds
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="column"
            >
              <div
                class="pokemon-column"
              >
                Ability:
              </div>
              <div
                class="pokemon-value"
              />
            </div>
            <div
              class="column"
            >
              <div
                class="pokemon-column"
              >
                Hidden Ability:
              </div>
              <div
                class="pokemon-value"
              >
                --
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card u-full-width fire"
        data-testid="card"
      >
        <div
          class="imageWrapper"
        >
          <img
            class="pokemon-image"
            src=""
          />
        </div>
        <div
          class="textWrapper"
        >
          <div
            class="pokemon-type"
          >
            fire
          </div>
          <div
            class="pokemon-name"
          >
            charizard
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              HP:
            </div>
             
            <div
              class="pokemon-value"
            >
              35
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Attack:
            </div>
             
            <div
              class="pokemon-value"
            >
              55
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Defense:
            </div>
             
            <div
              class="pokemon-value"
            >
              40
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Special Attack:
            </div>
             
            <div
              class="pokemon-value"
            >
              50
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Special Defense:
            </div>
             
            <div
              class="pokemon-value"
            >
              50
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Speed:
            </div>
             
            <div
              class="pokemon-value"
            >
              90 m/s
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Height:
            </div>
             
            <div
              class="pokemon-value"
            >
              undefined feets
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="pokemon-column"
            >
              Weight:
            </div>
             
            <div
              class="pokemon-value"
            >
              undefined pounds
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="column"
            >
              <div
                class="pokemon-column"
              >
                Ability:
              </div>
              <div
                class="pokemon-value"
              />
            </div>
            <div
              class="column"
            >
              <div
                class="pokemon-column"
              >
                Hidden Ability:
              </div>
              <div
                class="pokemon-value"
              >
                --
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
